Python/chapter03/p04_queue_via_stacks/miguelHx.py

```diff
@@ -0,0 +1,290 @@
+        Removes the top item from the stack
+        Raises:
+            IndexError: raised when pop is attempted on empty stack
+        Returns:
+            int: The data at the top of the stack
+        """
+        if self.top is None:
+            raise IndexError('Stack is Empty.')
+        item = self.top.data
+        self.top = self.top.next
+        self.current_node = self.top
+        self._size -= 1
+        return item
+
+    def push(self, item: T) -> None:
+        """
+        Adds an item to the top of the stack
+        Args:
+            item (int): data we want at the top of stack
+        """
+        t = StackNode(item, None)
+        t.next = self.top
+        self.top = t
+        self.current_node = self.top
+        self._size += 1
+
+    def peek(self) -> T:
+        """
+        Returns data at the top of the stack
+        Raises:
+            IndexError: [description]
+        Returns:
+            int: the value at the top of the stack
+        """
+        if self.top is None:
+            raise IndexError('Stack is Empty')
+        return self.top.data
+    
+    def empty(self) -> None:
+        while self._size > 0:
+            self.pop()
+
+    def __iter__(self) -> Iterator:
+        """
+        Builds a list of the current stack state.
+        For example, given the following stack:
+            3 -> 2 -> 1, where 3 is the top,
+        Expect:
+            [3, 2, 1]
+        Returns:
+            List[int]: list of integers
+        """
+        return self
+
+    def __next__(self) -> T:
+        self.index += 1
+        if self.index == self._size or self.current_node is None:
+            self.index = -1
+            self.current_node = self.top
+            raise StopIteration
+        n: T = self.current_node.data
+        self.current_node = self.current_node.next
+        return n
+
+    def __bool__(self) -> bool:
+        """
+        True is returned when the container is not empty.
+        From https://docs.python.org/3/reference/datamodel.html#object.__bool__ :
+        Called to implement truth value testing and the built-in operation bool();
+        should return False or True. When this method is not defined, len() is called,
+        if it is defined, and the object is considered true if its result is nonzero.
+        If a class defines neither len() nor bool(), all its instances are considered true.
+        Returns:
+            bool: False when empty, True otherwise
+        """
+        return self._size > 0
+    
+    def __len__(self) -> int:
+        return self._size
+
+    def __str__(self):
+        if self._size == 0:
+            return '<Empty>'
+        values = []
+        n = self.top
+        while n.next is not None:
+            values.append(str(n.data))
+            n = n.next
+        values.append(str(n.data))
+        return '->'.join(values)
+
+
+@dataclass
+class QueueNode(Generic[T]):
+    data: T
+    next: 'Optional[QueueNode[T]]'
+
+
+class MyQueue:
```

Comment:
```suggestion
class MyQueue(Generic[T]):
```

this way it's legal to refer to `T` inside this class.

Reply:
Updated here 4e38eabef4b023618802ec82779d57318bf16a47